src/fides/api/util/tcf/tc_string.py

```diff
@@ -0,0 +1,182 @@
+
+
+def build_tc_string(model: TCModel) -> str:
+    """Construct a TC String from the given TCModel
+
+    Currently only core and vendors_disclosed sections are supported.
+    """
+    core_string: str = build_core_string(model)
+    vendors_disclosed_string: str = build_disclosed_vendors_string(model)
+
+    return core_string + "." + vendors_disclosed_string
+
+
+def build_core_string(model: TCModel) -> str:
+    """
+    Build the "core" TC String
+
+    https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20Consent%20string%20and%20vendor%20list%20formats%20v2.md#the-core-string
+    """
+    max_vendor_consents: int = _get_max_vendor_id(model.vendor_consents)
+    max_vendor_li: int = _get_max_vendor_id(model.vendor_legitimate_interests)
+
+    # List of core fields.  Order is intentional!
+    core_fields: list = [
+        TCField(name="version", bits=6),
+        TCField(name="created", bits=36),
+        TCField(name="last_updated", bits=36),
+        TCField(name="cmp_id", bits=12),
+        TCField(name="cmp_version", bits=12),
+        TCField(name="consent_screen", bits=6),
+        TCField(name="consent_language", bits=12),
+        TCField(name="vendor_list_version", bits=12),
+        TCField(name="policy_version", bits=6),
+        TCField(name="is_service_specific", bits=1),
+        TCField(name="use_non_standard_texts", bits=USE_NON_STANDARD_TEXT_BITS),
+        TCField(name="special_feature_optins", bits=SPECIAL_FEATURE_BITS),
+        TCField(name="purpose_consents", bits=PURPOSE_CONSENTS_BITS),
+        TCField(
+            name="purpose_legitimate_interests", bits=PURPOSE_LEGITIMATE_INTERESTS_BITS
+        ),
+        TCField(name="purpose_one_treatment", bits=1),
+        TCField(name="publisher_country_code", bits=12),
+        TCField(name="max_vendor_consent", bits=16, value_override=max_vendor_consents),
+        TCField(
+            name="is_vendor_consent_range_encoding", bits=1, value_override=0
+        ),  # Using bitfield
+        TCField(name="vendor_consents", bits=max_vendor_consents),
+        TCField(name="max_vendor_li", bits=16, value_override=max_vendor_li),
+        TCField(
+            name="is_vendor_li_range_encoding", bits=1, value_override=0
+        ),  # Using bitfield
+        TCField(name="vendor_legitimate_interests", bits=max_vendor_li),
+        TCField(name="num_pub_restrictions", bits=12),
+    ]
+
+    core_bits: str = get_bits_for_section(core_fields, model)
+    return base64.urlsafe_b64encode(_convert_bitstring_to_bytes(core_bits)).decode()
+
+
+def build_disclosed_vendors_string(model: TCModel) -> str:
+    """Build the Optional Disclosed Vendors" section of the TC String
+
+    https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20Consent%20string%20and%20vendor%20list%20formats%20v2.md#disclosed-vendors
```

Comment:
reading this makes me think maybe the FE isn't handling this part right. I think the lib automatically sets all vendors in the GVL to disclosed = true. but we should probably only be doing the vendors that we are showing in the data map

Reply:
oh actually taking a closer look I think you're doing the same thing since you're getting `internal_gvl_vendor_ids`. I wonder if that's the right approach though? (I know you're copying the FE, this is more of a product q and a discussion of if we are doing it wrong on the FE 😅 )

> The DisclosedVendors is an optional TC String segment that records which vendors have been disclosed to a given user by a CMP. It may be used by a CMP while storing TC Strings, but must not be included in the TC String when returned by the CMP API.

to me that reads like it should be the vendors that show up in the vendors tab?

Reply:
ah wonderful, I thought that was odd myself, but was just matching the library.  Let me update!

Reply:
OK @allisonking!  I think I've got it.

So for an accept-all string, for vendors 2, 8, and 46, the backend is now building:

```
CPytTYAPytTYAAMABBENATEEAPLAAEPAAAAAAEEEALgCAAAAAAgAAAAA.IAXEEAAAAABA
```

and the FE is building:

```
CPytTYAPytTYAAMABBENASEEAPLAAEPAAAAAAEEEALoAQACAAXAAAAAA.IgoMV_H3_bX9v8Tq16ft0eYFX99j77uQxBhfJs-4FyDvW_JwVn2EyNAUytqYKmRIEuzZBIQFsHJnURVChQogVryHsYkGMhTEAJ-BkgHMRa2dYCFxvmYtjEQIZ5_p_czf52R_9_dn-3czjT9lnnzVtO2dlcKidS59tH_l_bRKb-7Ied-7W_4v09tfrkW_eTVu_9evvp9-u_s_v__934AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQ
```

Love the length reduction too!

Reply:
nice, looks less like the string is screaming 😄 I'll make a ticket to update in the FE too 👍

Reply:
Thanks again for surfacing this, I probably would have just kept matching the library without your commen!t 🙏

Reply:
https://github.com/ethyca/fides/issues/4165